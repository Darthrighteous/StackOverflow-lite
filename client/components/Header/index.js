import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import ButtonPrimary from '../Common/ButtonPrimary';
import verifyToken from '../../utils/verifyToken';
import Toaster from '../../utils/Toaster';

/**
 * Header component
 */
class Header extends Component {
  constructor(props) {
    super(props);

    this.handleLogOut = this.handleLogOut.bind(this);
  }

  componentDidMount() {
    const { setLoggedIn } = this.props;
    if(verifyToken()) {
      setLoggedIn(true);
    }
  }

  handleLogOut() {
    const { setLoggedIn } = this.props;
    localStorage.removeItem('user');

    setLoggedIn(false);
    Toaster.success('successfully logged out', 'Goodbye');
  }

  render() {
    const { loginStatus } = this.props;
    const user = JSON.parse(localStorage.getItem('user'));
    return (
      <header>
        <div className="header-content">
          <div className="title-search-cont">
            <Link to="/">
              <div className="title">
                <i className="fab fa-stack-overflow" />
                <span>stackoverflow-lite</span>
              </div>
            </Link>
            
            <form className="search-form">
              <Link to="/search">
                <i className="fas fa-search" />
              </Link>
              <input
                id="search-bar"
                type="text"
                name="search"
                placeholder="Search all questions..."
              />
            </form>
          </div>

          <div className="user-options-cont">
            {(loginStatus) ? (
              <div className="user-options">
                <Link to={`/profile/${user.username}`}>
                  <i className="fas fa-user-circle" />
                </Link>
                <ButtonPrimary
                  label="Log Out"
                  colorTxt="#3275c3"
                  colorBtn="#ebf4ff"
                  onClick={this.handleLogOut}
                />
              </div>
              ) : (
                <div className="guest-options">
                  <ButtonPrimary
                    link="/login"
                    label="Log In"
                    colorTxt="#3275c3"
                    colorBtn="#ebf4ff"
                  />
                  <ButtonPrimary
                    link="/signup"
                    label="Sign Up"
                    colorTxt="#fff"
                    colorBtn="#3275c3"
                  />
                </div>
              )
            }
          </div>
        </div>
      </header>
    );
  }
}

Header.propTypes = {
  loginStatus: PropTypes.bool.isRequired,
  setLoggedIn: PropTypes.func.isRequired,
};

export default Header;
