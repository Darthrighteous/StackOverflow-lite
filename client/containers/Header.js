import { connect } from 'react-redux';

import Header from '../components/Header';

const mapStateToProps = state => ({
  loginStatus: state.global.isLoggedIn
});

export default connect(mapStateToProps, null, null, { pure: false })(Header);