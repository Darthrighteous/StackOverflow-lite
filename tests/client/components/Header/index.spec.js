import React from 'react';
import { shallow } from 'enzyme';

import Header from '../../../../client/components/Header';

beforeAll(() => {
  process.env = Object.assign(process.env, { SECRET: 'test' });
});

describe('Header snapshot tests', () => {

  const user = {
    answer_count: 5,
    comment_count: 8,
    downvoted_answers: [1, 6, 7, 7],
    downvoted_questions: [1, 2, 4],
    email: "ja.ogunniyi@test.com",
    firstname: "Jalil",
    id: 22,
    joined_at: "2018-12-26T08:24:09.197Z",
    lastname: "Ogunniyi",
    question_count: 1,
    upvoted_answers: [1],
    upvoted_questions: [2, 1, 4],
    username: "Darthrighteous",
    token: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwibmFtZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTE2MjM5MDIyfQ.5mhBHqs5_DTLdINd9p5m7ZJ6XD0Xc55kIaCRY5r6HRA',
  };

  afterEach(() => {
    localStorage.clear();
  });

  const props = {
    loginStatus: true,
    setLoggedIn: jest.fn(),
  };
  test(' - logged in', () => {
    localStorage.setItem('user', JSON.stringify(user));
    const headerWrapper = shallow(<Header {...props} />);
    expect(headerWrapper).toMatchSnapshot();

    headerWrapper.find('ButtonPrimary').simulate('click');
  });

  const loggedOutProps = {
    ...props,
    loginStatus: false
  };
  test('wrong token - logged out', () => {
    const wrongTokenUser = {...user, token: 'xdxdxd'};
    localStorage.setItem('user', JSON.stringify(wrongTokenUser));
    const headerWrapper = shallow(<Header {...loggedOutProps} />);
    expect(headerWrapper).toMatchSnapshot();
  });

  test('no user - logged out', () => {
    const headerWrapper = shallow(<Header {...loggedOutProps} />);
    expect(headerWrapper).toMatchSnapshot();
  });
});
