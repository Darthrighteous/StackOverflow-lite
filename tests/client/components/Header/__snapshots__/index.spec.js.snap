// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot tests  - logged in 1`] = `
<header>
  <div
    className="header-content"
  >
    <div
      className="title-search-cont"
    >
      <Link
        replace={false}
        to="/"
      >
        <div
          className="title"
        >
          <i
            className="fab fa-stack-overflow"
          />
          <span>
            stackoverflow-lite
          </span>
        </div>
      </Link>
      <form
        className="search-form"
      >
        <Link
          replace={false}
          to="/search"
        >
          <i
            className="fas fa-search"
          />
        </Link>
        <input
          id="search-bar"
          name="search"
          placeholder="Search all questions..."
          type="text"
        />
      </form>
    </div>
    <div
      className="user-options-cont"
    >
      <div
        className="user-options"
      >
        <Link
          replace={false}
          to="/profile/Darthrighteous"
        >
          <i
            className="fas fa-user-circle"
          />
        </Link>
        <ButtonPrimary
          colorBtn="#ebf4ff"
          colorTxt="#3275c3"
          label="Log Out"
          link={null}
          onClick={[Function]}
        />
      </div>
    </div>
  </div>
</header>
`;

exports[`Header snapshot tests no user - logged out 1`] = `
<header>
  <div
    className="header-content"
  >
    <div
      className="title-search-cont"
    >
      <Link
        replace={false}
        to="/"
      >
        <div
          className="title"
        >
          <i
            className="fab fa-stack-overflow"
          />
          <span>
            stackoverflow-lite
          </span>
        </div>
      </Link>
      <form
        className="search-form"
      >
        <Link
          replace={false}
          to="/search"
        >
          <i
            className="fas fa-search"
          />
        </Link>
        <input
          id="search-bar"
          name="search"
          placeholder="Search all questions..."
          type="text"
        />
      </form>
    </div>
    <div
      className="user-options-cont"
    >
      <div
        className="guest-options"
      >
        <ButtonPrimary
          colorBtn="#ebf4ff"
          colorTxt="#3275c3"
          label="Log In"
          link="/login"
          onClick={null}
        />
        <ButtonPrimary
          colorBtn="#3275c3"
          colorTxt="#fff"
          label="Sign Up"
          link="/signup"
          onClick={null}
        />
      </div>
    </div>
  </div>
</header>
`;

exports[`Header snapshot tests wrong token - logged out 1`] = `
<header>
  <div
    className="header-content"
  >
    <div
      className="title-search-cont"
    >
      <Link
        replace={false}
        to="/"
      >
        <div
          className="title"
        >
          <i
            className="fab fa-stack-overflow"
          />
          <span>
            stackoverflow-lite
          </span>
        </div>
      </Link>
      <form
        className="search-form"
      >
        <Link
          replace={false}
          to="/search"
        >
          <i
            className="fas fa-search"
          />
        </Link>
        <input
          id="search-bar"
          name="search"
          placeholder="Search all questions..."
          type="text"
        />
      </form>
    </div>
    <div
      className="user-options-cont"
    >
      <div
        className="guest-options"
      >
        <ButtonPrimary
          colorBtn="#ebf4ff"
          colorTxt="#3275c3"
          label="Log In"
          link="/login"
          onClick={null}
        />
        <ButtonPrimary
          colorBtn="#3275c3"
          colorTxt="#fff"
          label="Sign Up"
          link="/signup"
          onClick={null}
        />
      </div>
    </div>
  </div>
</header>
`;
